app/trace/trace_viewer.tsx

```diff
@@ -282,6 +339,134 @@ export default class TraceViewer extends React.Component<TraceViewProps, {}> {
     }
   }
 
+  // Callback for keydown events in the search input.
+  private onSearchKeyDown = (e: React.KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === "Enter") {
+      e.preventDefault();
+      this.scrollToNextMatch();
+    }
+  };
+
+  // Returns true if the given event matches the provided filter string.
+  private eventMatchesFilter(event: TraceEvent, filter: string): boolean {
+    if (!filter) return true;
+    const f = filter.toLowerCase();
+    return (
+      event.name.toLowerCase().includes(f) ||
+      event.cat.toLowerCase().includes(f) ||
+      event.args?.target?.toLowerCase().includes(f) ||
+      event.args?.mnemonic?.toLowerCase().includes(f) ||
+      event.out?.toLowerCase().includes(f)
+    );
+  }
+
+  // Scrolls to and highlights the next event that matches the current search filter.
+  private scrollToNextMatch() {
+    const filter = this.getFilter();
+    if (!filter) return;
+
+    // Find the next matching event after `this.searchIndex`.
+    const total = this.searchableEvents.length;
+    if (!total) return;
+
+    let nextIndex = this.searchIndex;
+    for (let i = 0; i < total; i++) {
+      nextIndex = (nextIndex + 1) % total;
```

Comment:
Can you add a `direction` parameter to this function, and reference it here instead of `+1`? Then in `onKeyDown`, if `e.shiftKey` is held, set `direction` to `-1`? That would allow navigating both forward and backward through the matches.